Wikipedia/assets/codemirror/codemirror-editTextSelection.js

```diff
@@ -0,0 +1,126 @@
+  getRestrictiveRegex(maxAdjacentWordsToUse) {
+    const atLeastOneNonWordPattern = '\\W+'
+    return this.regexForLocatingSelectedTextWithPatternForSpace(atLeastOneNonWordPattern, maxAdjacentWordsToUse)
+  }
+  
+  // Slower, more error prone - but has less trouble with skipping past some unexpectedly complicated wiki markup, so less prone to not finding any match.
+  getPermissiveRegex(maxAdjacentWordsToUse) {
+    const atLeastOneCharPattern = '.+'
+    return this.regexForLocatingSelectedTextWithPatternForSpace(atLeastOneCharPattern, maxAdjacentWordsToUse)
+  }
+    
+  regexForLocatingSelectedTextWithPatternForSpace(patternForSpace, maxAdjacentWordsToUse) {
+    const replaceSpaceWith = (s, replacement) => s.replace(/\s+/g, replacement)
+
+    // Keep only the last 'maxAdjacentWordsToUse' words of 'textBeforeSelectedText'
+    const shouldKeepWordBeforeSelection = (e, i, a) => (a.length - i - 1) < maxAdjacentWordsToUse
+    // Keep only the first 'maxAdjacentWordsToUse' words of 'textAfterSelectedText'
+    const shouldKeepWordAfterSelection = (e, i) => i < maxAdjacentWordsToUse
+    const wordsBefore = this.textBeforeSelectedText.split(' ').filter(shouldKeepWordBeforeSelection).join(' ')
+    const wordsAfter = this.textAfterSelectedText.split(' ').filter(shouldKeepWordAfterSelection).join(' ')
+
+    const selectedTextPattern = replaceSpaceWith(this.selectedText, patternForSpace)
+    const textBeforeSelectedTextPattern = replaceSpaceWith(wordsBefore, patternForSpace)
+    const textAfterSelectedTextPattern = replaceSpaceWith(wordsAfter, patternForSpace)
+
+    // Attempt to locate wikitext selection based on the non-wikitext context strings above.
+    const atLeastOneNonWordOrOptionalStringPattern = '(?:\\W+|.*)'
+    const pattern = `(${textBeforeSelectedTextPattern.length > 0 ? '.*?' : ''}${textBeforeSelectedTextPattern}${atLeastOneNonWordOrOptionalStringPattern})(${selectedTextPattern})${atLeastOneNonWordOrOptionalStringPattern}${textAfterSelectedTextPattern}`
+    const regex = new RegExp(pattern, 's')
+
+    return regex
+  }
+}
+
+const wikitextRangeForSelectedAndAdjacentText = (selectedAndAdjacentText, wikitext) => {
+  const regex = selectedAndAdjacentText.regexForLocatingSelectedTextInWikitext(wikitext)
+  if (regex === null) {
+    return null
+  }
+  const match = wikitext.match(regex)
+  const matchedWikitextBeforeSelection = match[1]
+  const matchedWikitextSelection = match[2]
+  const wikitextRange = getWikitextRangeToSelect(matchedWikitextBeforeSelection, matchedWikitextSelection)
+  return wikitextRange
+}
+
+const getWikitextRangeToSelect = (wikitextBeforeSelection, wikitextSelection) => {
+  const wikitextBeforeSelectionLines = wikitextBeforeSelection.split('\n')
+  const startLine = wikitextBeforeSelectionLines.length - 1 
+  const startCh = wikitextBeforeSelectionLines.pop().length
+
+  const wikitextSelectionLines = wikitextSelection.split('\n')
+  const endLine = startLine + wikitextSelectionLines.length - 1
```

Comment:
should we check if these strings are not `undefined` before we call `split`?

Reply:
just tested - now that we're checking the `match` for null it looks like these strings are never `undefined` here - they can be zero length, but that's sometimes expected and doesn't cause an issue with `split`, so... good to go i think :)